tests/Unit/Time/TimeSteppers/Test_DormandPrince5.cpp

```diff
@@ -0,0 +1,30 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "tests/Unit/TestingFramework.hpp"
+
+#include "Parallel/PupStlCpp11.hpp"
+#include "Time/TimeSteppers/DormandPrince5.hpp"
+#include "Time/TimeSteppers/TimeStepper.hpp"
+#include "tests/Unit/TestCreation.hpp"
+#include "tests/Unit/TestHelpers.hpp"
+#include "tests/Unit/Time/TimeSteppers/TimeStepperTestUtils.hpp"
+
+SPECTRE_TEST_CASE("Unit.Time.TimeSteppers.DormandPrince5", "[Unit][Time]") {
+  const TimeSteppers::DormandPrince5 stepper{};
+  TimeStepperTestUtils::check_substep_properties(stepper);
+  TimeStepperTestUtils::integrate_test(stepper, 0, 1.0, 1.0e-9);
+  TimeStepperTestUtils::integrate_test(stepper, 0, -1.0, 1.0e-9);
+  TimeStepperTestUtils::integrate_variable_test(stepper, 0, 1.0e-9);
+  TimeStepperTestUtils::check_convergence_order(stepper, 5);
+  TimeStepperTestUtils::stability_test(stepper);
+  // Note: For DP5, dense output with no additional function evaluations
+  // is fourth-order accurate.
+  TimeStepperTestUtils::check_dense_output(stepper, 4);
```

Comment:
Your dense output function is a quartic, so I would expect 5th-order accuracy (in GTE-style counting) for the polynomial coefficients to not be underdetermined.

Reply:
I agree, and I think my comment here is unclear if not wrong. Currently, the test passes for check_dense_output(stepper, 4) but fails for check_dense_output(stepper, 5). In the latter case, the order reported is about 4.2.

I'll check again to see if I made a mistake in the dense output code, but I now suspect the problem is not having the derivative at the final time. I approximated f(t+dt, u1) as l6...perhaps that's lowering the dense output accuracy?

Reply:
It could also be the numerical order-estimation failing.  I've found that task to be very difficult to automate well and the code is not super robust.  I'll take a quick look at that.

Reply:
Now I'm really confused...I tried adding a trivial extra substep, so I didn't have to approximate f(t+dt, u1), but this did not significantly change the order reported by check_dense_output(). I checked the code for dense output against Numerical Recipes and SpEC, and I don't see any difference.

I reread Numerical Recipes, and it claims O(dt^5) accuracy in dense output. Maybe there's a bug in my dense output code that I can't see for starting at it for too long?

Reply:
The convergence estimate is getting it right, although I'm not sure that isn't just luck.

But I agree that something odd is going on with the math here and maybe the derivative approximation isn't good.  I'm going to look at that a bit more.

Reply:
Oops...I posted my last message before seeing your reply that the order-estimation might be having trouble. Thanks @wthrowe for taking a look into this!

Reply:
You can't get higher than 4th order dense output without additional function evaluations. This is stated in NR and also in the Hairer et al. book on solving ODEs.

Reply:
I know, but I thought what Nunerical Recipes falls 4th order is O(dt^5) errors, which would mean check_demse_output(..., 5). Is that not right?

Reply:
That would mean that you should be passing `6` to `check stepper order`, no? Assuming the dense output and stepper order functions use the same definition of order the dense output should be one less than the stepper order.

Reply:
Maybe you're right, @nilsdeppe! Everywhere else, when we put in a number for order, it's the order of the accuracy, not the order of the error term (RK3 puts a 3 in for order, and RK4 puts in a 4 for the order). So maybe this was right as I originally coded it after all? @wthrowe, do you agree, or is there still something strange going on here?

Reply:
@geoffrey4444 , the full accuracy of DP5 and the accompanying dense output routine uses the FSAL trick (first-same-as-last). In other words, a 7th derivative term is computed at f(x_n + h, y_{n+1}) at the end of the step. (Here f is the rhs.) This is then re-used as the 1st derivative term in the next step. I'm busy packing for CA, so didn't have time to check if you were doing this or not.

Reply:
I’m not...I wanted to get the base version done and then add FSAL later. But I did try adding a seventh step, so I could have the full results of u and du/dt at t+dt for dense output, but this made no difference in the test’s reported order for dense output. I might have misunderstood, but I thought FSAL just affected speed, not order?

Reply:
Yes, let's merge now and file an issue to re-examine FSAL. I think it does affect the order, and is not just a speed optimization. But I wouldn't bet on that!

Reply:
I've figured a few things out:
1. The order is reduced by using the last substep's derivative rather than the derivative at the end of the step.
2. There is a (unique) full-order dense output scheme that does not need the FSAL evaluation.  The FSAL scheme is better in that it is C1 at the step boundaries (obviously not possible without FSAL) (both are C0), and, at least for the test problem I was using, the coefficient of the error term is smaller.
3. The order estimator in the test fails to get the non-FSAL convergence rate correct.  I suspect it will do better with the FSAL one because there is less variation in the errors for the test problem used, (I'm guessing that the non-FSAL error changes sign so it is sometimes smaller than expected.) but it really seems to be working largely by luck and I need to write a more robust one.

For the record, here is the full-order non-FSAL dense output:
```c++
  const double coef2 =
      4.0 / 3339.0 * square(output_fraction) *
      (3162.0 + output_fraction * (-4682.0 + output_fraction * 1895.0));
  const double coef3 =
      -1.0 / 192.0 * square(output_fraction) *
      (324.0 + output_fraction * (-864.0 + output_fraction * 415));
  const double coef4 =
      -243.0 / 33920.0 * square(output_fraction) *
      (36.0 + output_fraction * (-176.0 + output_fraction * 185.0));
  const double coef5 =
      11.0 / 420.0 * square(output_fraction) *
      (36.0 + output_fraction * (-116.0 + output_fraction * 85.0));

  *u = history.begin().value() +
      dt * (
       (output_fraction - coef2 - coef3 - coef4 - coef5) *
           history.begin().derivative() +
       coef2 * (history.begin() + 2).derivative() +
       coef3 * (history.begin() + 3).derivative() +
       coef4 * (history.begin() + 4).derivative() +
       coef5 * (history.begin() + 5).derivative());
```